File: 04_schedule/problems/01_count_primes/count_primes.cpp
#include <cstdio>
#include <cstdlib>

int is_prime(long k) {
  if (k < 2) return 0;
  for (long d = 2; d * d <= k; d++) {
    if (k % d == 0) return 0;
  }
  return 1;
}

int main(int argc, char ** argv) {
  long N = (argc > 1 ? atol(argv[1]) : 300000L);
  long count = 0;
  // TODO: 下のループを #pragma omp parallel for schedule(runtime) reduction(+:count) で並列化せよ.
  for (long i = 2; i <= N; i++) {
    count += is_prime(i);
  }
  printf("number of primes <= %ld : %ld\n", N, count);
  return 0;
}

File: 04_schedule/problems/01_count_primes/count_primes.f90
module prime_mod
contains
  function is_prime(k) result(r)
    integer(8), intent(in) :: k
    integer :: r
    integer(8) :: d
    if (k < 2) then
       r = 0
       return
    end if
    d = 2
    do while (d * d <= k)
       if (mod(k, d) == 0) then
          r = 0
          return
       end if
       d = d + 1
    end do
    r = 1
  end function is_prime
end module prime_mod

program count_primes
  use prime_mod
  character(len=64) :: arg
  integer(8) :: N, i, count
  N = 300000_8
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) N
  end if
  count = 0
  ! TODO: 下のループを !$omp parallel do schedule(runtime) reduction(+:count) で並列化せよ.
  do i = 2, N
     count = count + is_prime(i)
  end do
  ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
  print "(a,i0,a,i0)", "number of primes <= ", N, " : ", count
end program count_primes
